docs/azure/sdk/unit-testing-mocking.md

```diff
@@ -0,0 +1,420 @@
+* All public client members are virtual to allow overriding.
+
+> [!NOTE]
+> The code examples in this article use types from the [`Azure.Security.KeyVault.Secrets`](https://www.nuget.org/packages/Azure.Security.KeyVault.Secrets/) library for the Azure Key Vault service. The concepts demonstrated in this article also apply to service clients from many other Azure services, such as Azure Storage or Azure Service Bus.
+
+To create a test service client, you can either use a mocking library such as [Moq](https://www.nuget.org/packages/moq/) or standard C# features such as inheritance. Mocking frameworks allow you to simplify the code that you must write to override member behavior (as well as other useful features that are beyond the scope of this article).
+
+# [Moq](#tab/moq)
+
+```csharp
+KeyVaultSecret keyVaultSecret = SecretModelFactory.KeyVaultSecret(
+    new SecretProperties("secret"), "secretValue");
+Mock<SecretClient> clientMock = new Mock<SecretClient>();
+clientMock.Setup(c => c.GetSecret(
+        It.IsAny<string>(),
+        It.IsAny<string>(),
+        It.IsAny<CancellationToken>())
+    )
+    .Returns(Response.FromValue(keyVaultSecret, Mock.Of<Response>()));
+
+clientMock.Setup(c => c.GetSecretAsync(
+        It.IsAny<string>(), 
+        It.IsAny<string>(), 
+        It.IsAny<CancellationToken>())
+    )
+    .ReturnsAsync(Response.FromValue(keyVaultSecret, Mock.Of<Response>()));
+
+SecretClient secretClient = clientMock.Object;
+```
+
+# [C#](#tab/csharp)
+
+To create a test client instance using C# without a library, inherit from the client type and override methods you are calling in your code with an implementation that returns a set of test objects. Most clients contain both synchronous and asynchronous methods for operations; override only the one your application code is calling.
+
+> [!NOTE]
+> It can be cumbersome to manually define test classes, especially if you need to customize behavior differently for each test. Consider using a library like Moq to streamline your testing.
+
+```csharp
+public class MockSecretClient : SecretClient 
+{
+    public override Response<KeyVaultSecret> GetSecret(
+        string name,
+        string version = null, 
+        CancellationToken cancellationToken = default)
+        => throw new NotImplementedException();
+        
+    public override Task<Response<KeyVaultSecret>> GetSecretAsync(
+        string name, 
+        string version = null, 
+        CancellationToken cancellationToken = default)
+        => throw new NotImplementedException();
+}
+
+SecretClient secretClient = new MockSecretClient();
+```
+
+---
+
+### Service client input and output models
+
+Model types hold the data being sent and received from Azure services. There are two main kinds of models:
+
+* **Input models** are intended to be created and passed as parameters to service methods by developers. They have one or more public constructors and writeable properties.
```

Comment:
Pure input/output models are the majority case, but we may want to mention "round-trip" models here as well.  These are returned by the service, modified, and used as input.